### PR TITLE
[BugFix] validate privilege before grant to role

### LINKED ISSUE #10414: [BUG] grant usage_priv on database/table to role can execute without error
### Steps to reproduce the behavior (Required)
create database db;
create role role1;
grant usage_priv on db.* to role 'role1';
grant usage_priv on db to role 'role1';

### Expected behavior (Required)
grant usage_priv on db.* to role 'role1';
ERROR 1064 (HY000): Db privilege can not contains global or resource or impersonate privileges: Usage_priv 

grant usage_priv on db to role 'role1';
ERROR 1064 (HY000): Db privilege can not contains global or resource or impersonate privileges: Usage_priv 

### Real behavior (Required)
grant usage_priv on db.* to role 'role1';
Query OK, 0 rows affected (0.01 sec)

grant usage_priv on db to role 'role1';
Query OK, 0 rows affected (0.01 sec)

### StarRocks version (Required)
version 2.3

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [x] bugfix
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #10414 

## Problem Summary(Required) ：

In the previous implementation, privileges will only be validated before granting to a user. 
Suppose we're granting a `USAGE` privilege on a database; since the `USAGE` privilege is only meaningful on resources like Spark, this action should end in error. Unfortunately, it will execute successfully when granted to a role.

```
grant usage_priv on db.* to role 'role1';
Query OK, 0 rows affected (0.01 sec)
grant usage_priv on db.* to user;
ERROR 1064 (HY000): Db privilege can not contains global or resource or impersonate privileges: Usage_priv
```

This PR validates privilege before adding to RoleManager or PrivTable.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 8 (87.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|7|8|87.50%|[842]|

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 1]
```diff
@@ -827,6 +827,26 @@ public void replayGrant(PrivInfo privInfo) {
     private void grantInternal(UserIdentity userIdent, String role, TablePattern tblPattern,
                                PrivBitSet privs, boolean errOnNonExist, boolean isReplay)
             throws DdlException {
+
+        // check privilege
+        switch (tblPattern.getPrivLevel()) {
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:832]: If the isReplay is true, you should tolerate this error to make sure compatible.

**Review by waittttting — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 9 / 10 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|9|10|90.00%|[843]|

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java (hunk 1 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java]: Good advice, I totally forget it.

**Review by gengjun-git — DISMISSED** [review]

**Review by waittttting — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 2]
```diff
@@ -827,6 +827,29 @@ public void replayGrant(PrivInfo privInfo) {
     private void grantInternal(UserIdentity userIdent, String role, TablePattern tblPattern,
                                PrivBitSet privs, boolean errOnNonExist, boolean isReplay)
             throws DdlException {
+
+        if (! isReplay) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:831]: Redundant space

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 3]
```diff
@@ -857,6 +880,21 @@ private void grantInternal(UserIdentity userIdent, String role, TablePattern tbl
 
     private void grantInternal(UserIdentity userIdent, String role, ResourcePattern resourcePattern, PrivBitSet privs,
                                boolean errOnNonExist, boolean isReplay) throws DdlException {
+        if (! isReplay) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:883]: Redundant space

File: fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java [hunk 4]
```diff
@@ -827,6 +827,29 @@ public void replayGrant(PrivInfo privInfo) {
...
+                case DATABASE:
+                    if (privs.containsNodePriv() || privs.containsResourcePriv() || privs.containsImpersonatePriv()) {
+                        throw new DdlException(
+                                "Db privilege can not contains global or resource or impersonate privileges: " + privs);
+                    }
+                    break;
+
+                case TABLE:
+                    if (privs.containsNodePriv() || privs.containsResourcePriv() || privs.containsImpersonatePriv()) {
+                        throw new DdlException(
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/mysql/privilege/Auth.java:843]:
```suggestion
                        throw new DdlException("Table does not contains {}", privs);
```
The `Table privilege can not contains global or resource or impersonate privileges` is misunderstanding.

**Review by chaoyli — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 9 / 10 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|9|10|90.00%|[842]|

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
@Mergifyio backport branch-2.4 branch-2.3
¡Muchas gracias!

**Pull request merged**